core/runtime.rs

```diff
@@ -1882,100 +1922,40 @@ impl JsRuntime {
+    let isolate = self.v8_isolate.as_mut().unwrap();
     // Now handle actual ops.
     {
-      let mut state = state_rc.borrow_mut();
+      let mut state = self.state.borrow_mut();
       state.have_unpolled_ops = false;
-
       while let Poll::Ready(Some(item)) = state.pending_ops.poll_next_unpin(cx)
       {
-        let (context, promise_id, op_id, resp) = item;
+        let (context, resolver, promise_id, op_id, resp) = item;
         state.op_state.borrow().tracker.track_async_completed(op_id);
-        for (context2, results) in results_per_realm.iter_mut() {
-          if context == *context2 {
-            results.push((promise_id, resp));
-            break;
-          }
-        }
-        JsRealm::new(context)
-          .state(self.v8_isolate())
+
+        let realm = JsRealm::new(context);
+        realm
+          .state(isolate)
           .borrow_mut()
           .unrefed_ops
           .remove(&promise_id);
-      }
-    }
-
-    for (context, results) in results_per_realm {
-      if results.is_empty() {
-        continue;
-      }
-
-      let realm = JsRealm::new(context);
-      let js_recv_cb_handle = realm
-        .state(self.v8_isolate())
-        .borrow()
-        .js_recv_cb
-        .clone()
-        .unwrap();
-      let scope = &mut realm.handle_scope(self.v8_isolate());
 
-      // We return async responses to JS in unbounded batches (may change),
-      // each batch is a flat vector of tuples:
-      // `[promise_id1, op_result1, promise_id2, op_result2, ...]`
-      // promise_id is a simple integer, op_result is an ops::OpResult
-      // which contains a value OR an error, encoded as a tuple.
-      // This batch is received in JS via the special `arguments` variable
-      // and then each tuple is used to resolve or reject promises
-      let mut args = vec![];
-
-      for (promise_id, mut resp) in results.into_iter() {
-        args.push(v8::Integer::new(scope, promise_id).into());
-        args.push(match resp.to_v8(scope) {
-          Ok(v) => v,
-          Err(e) => OpResult::Err(OpError::new(&|_| "TypeError", e.into()))
-            .to_v8(scope)
-            .unwrap(),
-        });
-      }
-
-      let tc_scope = &mut v8::TryCatch::new(scope);
-      let js_recv_cb = js_recv_cb_handle.open(tc_scope);
-      let this = v8::undefined(tc_scope).into();
-      js_recv_cb.call(tc_scope, this, args.as_slice());
-
-      if let Some(exception) = tc_scope.exception() {
-        return exception_to_err_result(tc_scope, exception, false);
+        results.push((realm, resolver, resp));
       }
     }
 
+    for (realm, resolver, mut resp) in results {
+      let scope = &mut realm.handle_scope(isolate);
+      let resolver = resolver.open(scope);
+      let result = match resp.to_v8(scope) {
+        Ok(v) => v,
+        Err(e) => OpResult::Err(OpError::new(&|_| "TypeError", e.into()))
+          .to_v8(scope)
+          .unwrap(),
+      };
+      resolver.resolve(scope, result);
```

Comment:
Use reject on error?

Reply:
^ We have `crate::error::to_v8_error(scope, op_state.get_error_class_fn, &e)`